tests/qat/test_qatconfig.py

```diff
@@ -1,12 +1,30 @@
 # SPDX-License-Identifier: BSD-3-Clause
 # Copyright (c) 2023 Oxford Quantum Circuits Ltd
 import pytest
+from pydantic import ValidationError
 
 from qat import qatconfig
 
+MAX_REPEATS_LIMIT = 100_000
+qatconfig.MAX_REPEATS_LIMIT = MAX_REPEATS_LIMIT
 
-@pytest.mark.parametrize("repeat_limits", [10_000, 50_000, 16_874, 100_000])
-def test_changed_max_repeats_limit(repeat_limits):
-    # Default value is 100_000. Test manual change of repeats limit.
-    qatconfig.MAX_REPEATS_LIMIT = repeat_limits
-    assert qatconfig.MAX_REPEATS_LIMIT == repeat_limits
+
+@pytest.mark.parametrize("invalid_argument", ["invalid", {"key": 5}, 5.5])
+def test_qatconfig_invalid_assignment(invalid_argument):
+    with pytest.raises(ValidationError):
+        qatconfig.MAX_REPEATS_LIMIT = invalid_argument
+
+
+@pytest.mark.parametrize("repeats_limit", [None, 10_000, 16_874, 50_000, 100_000])
+def test_change_max_repeats_limit(repeats_limit):
+    qatconfig.__init__()  # Reload settings.
+    # Test direct change of repeats limit.
+    qatconfig.MAX_REPEATS_LIMIT = repeats_limit
+    assert qatconfig.MAX_REPEATS_LIMIT == repeats_limit
+
+
+@pytest.mark.parametrize("repeats_limit", [None, 10_000, 16_874, 50_000, 100_000])
+def test_change_env_max_repeats_limit(monkeypatch, repeats_limit):
+    monkeypatch.setenv("QAT_MAX_REPEATS_LIMIT", str(repeats_limit))
+    # Default value cannot change if env variable changes after instantiation.
+    assert qatconfig.MAX_REPEATS_LIMIT == MAX_REPEATS_LIMIT
```

Comment:
I think we need a test that setting via the env does work "at the start"...

```
from qat.purr.qatconfig import QatConfig
def test_set_via_env(monkeypatch):
    NEW_LIMIT = 15321
    monkeypatch.setenv("QAT_MAX_REPEATS_LIMIT", str(NEW_LIMIT))
    newconfig = QatConfig()
    assert newconfig.MAX_REPEATS_LIMIT == NEW_LIMIT
   ```
   
   something like that?

Reply:
Agreed, and additionally check if the existing `qatconfig` is not changed.

This, together with enforced values for `MAX_REPEATS_LIMIT`, is implemented in the latest commit.